<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.251">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Collection of helper functions to facilicate plotting">

<title>samviz - util</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="samviz - util">
<meta property="og:description" content="Collection of helper functions to facilicate plotting">
<meta property="og:site-name" content="samviz">
<meta name="twitter:title" content="samviz - util">
<meta name="twitter:description" content="Collection of helper functions to facilicate plotting">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">samviz</span>
    </a>
  </div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">util</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">samap-viz</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./util.html" class="sidebar-item-text sidebar-link active">util</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./plot.html" class="sidebar-item-text sidebar-link">plot</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#procrustes" id="toc-procrustes" class="nav-link active" data-scroll-target="#procrustes">procrustes</a></li>
  <li><a href="#grouped_obs_mean" id="toc-grouped_obs_mean" class="nav-link" data-scroll-target="#grouped_obs_mean">grouped_obs_mean</a></li>
  <li><a href="#grouped_obs_present" id="toc-grouped_obs_present" class="nav-link" data-scroll-target="#grouped_obs_present">grouped_obs_present</a></li>
  <li><a href="#grouped_obs_percent" id="toc-grouped_obs_percent" class="nav-link" data-scroll-target="#grouped_obs_percent">grouped_obs_percent</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/galicae/samap-viz/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">util</h1>
</div>

<div>
  <div class="description">
    Collection of helper functions to facilicate plotting
  </div>
</div>


<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>Functions in this module depend on <code>pandas</code>, <code>anndata</code>, and <code>scanpy</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/galicae/samap-viz/blob/main/samap_viz/util.py#L7" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="procrustes" class="level3">
<h3 class="anchored" data-anchor-id="procrustes">procrustes</h3>
<blockquote class="blockquote">
<pre><code> procrustes (x:str, appropriate_length:int=50, pad_with:str=' ',
             side:str='right')</code></pre>
</blockquote>
<p>A function to regulate string length.</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>x</td>
<td>str</td>
<td></td>
<td>input string</td>
</tr>
<tr class="even">
<td>appropriate_length</td>
<td>int</td>
<td>50</td>
<td>desired length</td>
</tr>
<tr class="odd">
<td>pad_with</td>
<td>str</td>
<td></td>
<td>character to pad with</td>
</tr>
<tr class="even">
<td>side</td>
<td>str</td>
<td>right</td>
<td>which side to pad on (“left”, “right”)</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>str</strong></td>
<td></td>
<td><strong>string with desired length</strong></td>
</tr>
</tbody>
</table>
<p>We are primarily going to be working with non-model species, so the gene names will always be of the form</p>
<p><code>XLOC_123456 | emapper-name-or-description-if-we're-lucky</code></p>
<p>or something similar. This means that we could have extreme variation in the actual length of a gene “name”; this will make it very hard to put gene names on axes as it will distort figure sizes. I wrote a function to either trim or pad strings; even though axis labels are not in monospace fonts, it is much easier to visually reconcile strings with lengths in the same order of magnitude.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>too_short <span class="op">=</span> <span class="st">'Niko'</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>just_right <span class="op">=</span> <span class="st">'Theseus'</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>too_tall <span class="op">=</span> <span class="st">'The Mountain'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> procrustes(just_right, appropriate_length<span class="op">=</span><span class="dv">7</span>) <span class="op">==</span> <span class="st">'Theseus'</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> procrustes(too_short, appropriate_length<span class="op">=</span><span class="dv">7</span>) <span class="op">==</span> <span class="st">'Niko   '</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> procrustes(too_tall, appropriate_length<span class="op">=</span><span class="dv">7</span>) <span class="op">==</span> <span class="st">'The Mou'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a id="grouped_obs_mean"></a></p>
<hr>
<p><a href="https://github.com/galicae/samap-viz/blob/main/samap_viz/util.py#L27" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="grouped_obs_mean" class="level3">
<h3 class="anchored" data-anchor-id="grouped_obs_mean">grouped_obs_mean</h3>
<blockquote class="blockquote">
<pre><code> grouped_obs_mean (adata:anndata._core.anndata.AnnData, group_key:str,
                   layer:str=None)</code></pre>
</blockquote>
<p>Helper function to calculate average expression per group in an <code>AnnData</code> object.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td></td>
<td>AnnData object to analyse</td>
</tr>
<tr class="even">
<td>group_key</td>
<td>str</td>
<td></td>
<td><code>.obs</code> category to group by</td>
</tr>
<tr class="odd">
<td>layer</td>
<td>str</td>
<td>None</td>
<td>layer to use. If none, use <code>.X</code></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>a groups<span class="math inline">\(\times\)</span>genes dataframe with the average expression</strong></td>
</tr>
</tbody>
</table>
<p>Many tasks in single-cell analysis require us to know the average expression of a gene in a certain group of cells. While <code>scanpy</code> <em>does</em> perform that task behind the scenes for, e.g.&nbsp;dotplots, this is not functionality that is exposed to the users. This is an implementation based on <a href="https://github.com/theislab/scanpy/issues/181#issuecomment-534867254">ivirshup’s answer</a> to a scanpy issue.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">'../example_data/hydra.h5ad'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>cluster_means <span class="op">=</span> grouped_obs_mean(adata, group_key<span class="op">=</span><span class="st">'Cluster'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If <span class="math inline">\(G\)</span> is the number of genes and <span class="math inline">\(C\)</span> the number of unique clusters in the <code>group_key</code>, the returned array should have the shape <span class="math inline">\(G \times C\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>no_genes <span class="op">=</span> adata.shape[<span class="dv">1</span>]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>no_clusters <span class="op">=</span> <span class="bu">len</span>(np.unique(adata.obs[<span class="st">'Cluster'</span>]))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> cluster_means.shape <span class="op">==</span> (no_genes, no_clusters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Additionally, each column of the array should contain the average detected expression for cells in that cluster:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>belong_to_ecEp_SC2 <span class="op">=</span> adata.obs[<span class="st">'Cluster'</span>] <span class="op">==</span> <span class="st">'ecEp_SC2'</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>ecEp_SC2_average <span class="op">=</span> np.mean(adata[belong_to_ecEp_SC2].X, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ecEp_SC2_average <span class="op">=</span> np.array(ecEp_SC2_average)[<span class="dv">0</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">all</span>(np.isclose(cluster_means[<span class="st">'ecEp_SC2'</span>], ecEp_SC2_average))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/galicae/samap-viz/blob/main/samap_viz/util.py#L50" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="grouped_obs_present" class="level3">
<h3 class="anchored" data-anchor-id="grouped_obs_present">grouped_obs_present</h3>
<blockquote class="blockquote">
<pre><code> grouped_obs_present (adata:anndata._core.anndata.AnnData, group_key:str,
                      layer:str=None)</code></pre>
</blockquote>
<p>Helper function to calculate how many cells express each gene per group in an <code>AnnData</code> object.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td></td>
<td>AnnData object to analyse</td>
</tr>
<tr class="even">
<td>group_key</td>
<td>str</td>
<td></td>
<td><code>.obs</code> category to group by</td>
</tr>
<tr class="odd">
<td>layer</td>
<td>str</td>
<td>None</td>
<td>layer to use. If none, use <code>.X</code></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>a groups<span class="math inline">\(\times\)</span>genes dataframe with the number of expressing cells</strong></td>
</tr>
</tbody>
</table>
<p>Another critical value to know when making dotplots is the fraction of cells expressing a gene in a certain cluster. Again, <code>scanpy</code> performs that task without exposing it to the users. Similar to <a href="#grouped_obs_mean">[<code>grouped_obs_mean</code>](https://galicae.github.io/samap-viz/util.html#grouped_obs_mean)</a> this is an implementation based on <a href="https://github.com/theislab/scanpy/issues/181#issuecomment-534867254">ivirshup’s answer</a> to a scanpy issue. Here we calculate the sum of cells expressing a gene, a table we can use to calculate the fraction later.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>num_expressing <span class="op">=</span> grouped_obs_present(adata, group_key<span class="op">=</span><span class="st">'Cluster'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If <span class="math inline">\(G\)</span> is the number of genes and <span class="math inline">\(C\)</span> the number of unique clusters in the <code>group_key</code>, the returned array should have the shape <span class="math inline">\(G \times C\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> num_expressing.shape <span class="op">==</span> (no_genes, no_clusters)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Additionally, each column of the array should contain the percentage of cells expressing each gene in that cluster:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>belong_to_ecEp_SC2 <span class="op">=</span> adata.obs[<span class="st">'Cluster'</span>] <span class="op">==</span> <span class="st">'ecEp_SC2'</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>ecEp_SC2_expr <span class="op">=</span> np.<span class="bu">sum</span>(adata[belong_to_ecEp_SC2].X<span class="op">&gt;</span><span class="dv">0</span>, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>ecEp_SC2_expr <span class="op">=</span> np.array(ecEp_SC2_expr)[<span class="dv">0</span>]</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">all</span>(num_expressing[<span class="st">'ecEp_SC2'</span>] <span class="op">==</span> ecEp_SC2_expr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/galicae/samap-viz/blob/main/samap_viz/util.py#L73" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="grouped_obs_percent" class="level3">
<h3 class="anchored" data-anchor-id="grouped_obs_percent">grouped_obs_percent</h3>
<blockquote class="blockquote">
<pre><code> grouped_obs_percent (adata:anndata._core.anndata.AnnData, group_key:str,
                      layer:str=None)</code></pre>
</blockquote>
<p>Helper function to calculate how many cells express each gene per group in an <code>AnnData</code> object.</p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>adata</td>
<td>AnnData</td>
<td></td>
<td>AnnData object to analyse</td>
</tr>
<tr class="even">
<td>group_key</td>
<td>str</td>
<td></td>
<td><code>.obs</code> category to group by</td>
</tr>
<tr class="odd">
<td>layer</td>
<td>str</td>
<td>None</td>
<td>layer to use. If none, use <code>.X</code></td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>DataFrame</strong></td>
<td></td>
<td><strong>a groups<span class="math inline">\(\times\)</span>genes dataframe with the number of expressing cells</strong></td>
</tr>
</tbody>
</table>
<p>Calculating the fraction of cells is of course very straightforward once we have counted the number of cells that express the gene as well as the total number of cells in a cluster.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>frac_expressing <span class="op">=</span> grouped_obs_percent(adata, group_key<span class="op">=</span><span class="st">'Cluster'</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use the counts and number of cells we calculated before</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>frac_ecEp_SC2 <span class="op">=</span> ecEp_SC2_expr <span class="op">/</span> np.<span class="bu">sum</span>(belong_to_ecEp_SC2)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">all</span>(np.isclose(frac_expressing[<span class="st">'ecEp_SC2'</span>], frac_ecEp_SC2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>